erizo/src/erizo/rtp/RtpPaddingGeneratorHandler.cpp

```diff
@@ -0,0 +1,217 @@
+#include "rtp/RtpPaddingGeneratorHandler.h"
+
+#include <algorithm>
+#include <string>
+
+#include "./MediaDefinitions.h"
+#include "./WebRtcConnection.h"
+#include "./RtpUtils.h"
+
+namespace erizo {
+
+DEFINE_LOGGER(RtpPaddingGeneratorHandler, "rtp.RtpPaddingGeneratorHandler");
+
+constexpr duration kStatsPeriod = std::chrono::milliseconds(100);
+constexpr duration kFastStartMaxDuration = std::chrono::seconds(30);
+constexpr uint8_t kMaxPaddingSize = 255;
+constexpr uint64_t kMinMarkerRate = 3;
+constexpr uint8_t kMaxFractionLostAllowed = .05 * 255;  // 5% of packet losts
+constexpr uint64_t kInitialRembValue = 300000;
+
+RtpPaddingGeneratorHandler::RtpPaddingGeneratorHandler(std::shared_ptr<erizo::Clock> the_clock) :
+  clock_{the_clock}, connection_{nullptr}, max_video_bw_{0}, higher_sequence_number_{0},
+  video_sink_ssrc_{0}, audio_source_ssrc_{0},
+  number_of_full_padding_packets_{0}, last_padding_packet_size_{0},
+  last_rate_calculation_time_{clock_->now()}, started_at_{clock_->now()},
+  enabled_{false}, first_packet_received_{false},
+  marker_rate_{std::chrono::milliseconds(100), 20, 1., clock_},
+  padding_bitrate_{std::chrono::milliseconds(100), 10, 8., clock_},
+  rtp_header_length_{12}, remb_value_{kInitialRembValue}, fast_start_{true} {}
+
+
+void RtpPaddingGeneratorHandler::enable() {
+}
+
+void RtpPaddingGeneratorHandler::disable() {
+}
+
+void RtpPaddingGeneratorHandler::notifyUpdate() {
+  auto pipeline = getContext()->getPipelineShared();
+  if (pipeline && !connection_) {
+    connection_ = pipeline->getService<WebRtcConnection>().get();
+    video_sink_ssrc_ = connection_->getVideoSinkSSRC();
+    audio_source_ssrc_ = connection_->getAudioSinkSSRC();
+    stats_ = pipeline->getService<Stats>();
+  }
+
+  auto quality_manager = pipeline->getService<QualityManager>();
+
+  if (quality_manager->isPaddingEnabled() && !enabled_) {
+    enablePadding();
+  } else if (!quality_manager->isPaddingEnabled() && enabled_) {
+    disablePadding();
+  }
+
+  auto processor = pipeline->getService<RtcpProcessor>();
+  if (processor) {
+    max_video_bw_ = processor->getMaxVideoBW();
+  }
+}
+
+void RtpPaddingGeneratorHandler::read(Context *ctx, std::shared_ptr<dataPacket> packet) {
+  RtpUtils::forEachRRBlock(packet, [this](RtcpHeader *chead) {
```

Comment:
I'm thinking we should move as much as we can to the stats. Here it makes sense to react as soon as we get the info but I keep thinking we should avoid parsing the same packets n times.